### PR TITLE
[NEW][8.0] website_sale_checkout_country_vat: New module

### PR DESCRIPTION
**Pull request by sergio-teruel**
New module to autocomplete VAT code in website checkout form. This module depends from website/website_snippet_country_dropdown but it's in other PR (https://github.com/OCA/website/pull/210).
This PR close https://github.com/OCA/e-commerce/pull/109

@Tecnativa
@carlosdauden please review.

### CODE REVIEW

**Comment by rafaelbn** [comment]
This is a WIP until https://github.com/OCA/website/pull/210 get merged. Thanks

File: website_sale_checkout_country_vat/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
+# © 2016 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+{
+    "name": "Website Sale Checkout Country VAT",
+    "summary": "Autocomplete VAT in checkout process",
+    "version": "8.0.1.0.0",
+    "category": "e-commerce",
```
Comment (yajo) [inline website_sale_checkout_country_vat/__openerp__.py]: Category "Website", just like website_sale

File: website_sale_checkout_country_vat/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import http
+from openerp.http import request
+from openerp.addons.website_sale.controllers.main import website_sale
+
+
+class WebsiteSale(website_sale):
+
+    def _get_vat_code(self, code=None):
```
Comment (yajo) [inline website_sale_checkout_country_vat/controllers/main.py]: This method's name is misleading. You are not actually getting a code, but a country.

File: website_sale_checkout_country_vat/controllers/main.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        if code:
+            return request.env['res.country'].search([
+                ('code', '=', code)
+            ])
+        return (request.website.vat_code_id or
+                request.website.company_id.country_id)
+
+    @http.route(
+        ['/shop/confirm_order'], type='http', auth="public", website=True)
```
Comment (yajo) [inline website_sale_checkout_country_vat/controllers/main.py]: No need to pass all parameters to `route` if the method was already routed in parent controller. Just put `@http.route()` and all is inherited.

File: website_sale_checkout_country_vat/controllers/main.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    def confirm_order(self, **post):
+        if 'vat' in post:
+            if post['country_code_field'] != post['vat'][:2]:
```
Comment (yajo) [inline website_sale_checkout_country_vat/controllers/main.py]: Is this even possible?

File: website_sale_checkout_country_vat/controllers/main.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                post['vat'] = '%s%s' % (
+                    post['country_code_field'], post['vat'])
+        return super(WebsiteSale, self).confirm_order(**post)
+
+    def checkout_values(self, data=None):
+        res = super(WebsiteSale, self).checkout_values(data)
+        if 'checkout' in res:
+            vat = res['checkout'].get('vat', '')
+            if 'country_code_field' not in res['checkout']:
+                res['checkout']['country_code_field'] = vat[:2]
```
Comment (yajo) [inline website_sale_checkout_country_vat/controllers/main.py]: Better use [`setdefault`](https://docs.python.org/2.7/library/stdtypes.html#dict.setdefault), it's there for this purpose.

File: website_sale_checkout_country_vat/models/res_config.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+# © 2016 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import fields, models
+
+
+class WebsiteConfigSettings(models.TransientModel):
+    _inherit = 'website.config.settings'
+
+    vat_code_id = fields.Many2one(
```
Comment (yajo) [inline website_sale_checkout_country_vat/models/res_config.py]: Another misleading name. Better `default_vat_country_id`.

File: website_sale_checkout_country_vat/models/res_config.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+        related='website_id.vat_code_id', string='Default Country VAT Code')
```
Comment (yajo) [inline website_sale_checkout_country_vat/models/res_config.py]: "Default country for VAT codes"

File: website_sale_checkout_country_vat/models/website.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+# © 2016 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import fields, models
+
+
+class Website(models.Model):
+    _inherit = 'website'
+
+    vat_code_id = fields.Many2one(
```
Comment (yajo) [inline website_sale_checkout_country_vat/models/website.py]: Better default_vat_country_id.

File: website_sale_checkout_country_vat/models/website.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+        comodel_name='res.country', string="Default Country VAT Code", size=2)
```
Comment (yajo) [inline website_sale_checkout_country_vat/models/website.py]: "Default country for VAT codes"

File: website_sale_checkout_country_vat/static/src/css/style.css [hunk 10]
```diff
@@ -0,0 +1,6 @@
+/* © 2016 Sergio Teruel<sergio.teruel@recnativa.com>
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). */
```
Comment (yajo) [inline website_sale_checkout_country_vat/static/src/css/style.css]: LGPL

File: website_sale_checkout_country_vat/static/src/js/website_sale_checkout_country_vat.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+/* © 2016 Tecnativa, S.L.
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
```
Comment (yajo) [inline website_sale_checkout_country_vat/static/src/js/website_sale_checkout_country_vat.js]: LGPL

File: website_sale_checkout_country_vat/static/src/js/website_sale_checkout_country_vat.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+ */
+'use strict';
+(function ($) {
+    var website = openerp.website,
+        qweb = openerp.qweb;
+
+    website.snippet.animationRegistry.checkoutVat = website.snippet.animationRegistry.countryDropdown.extend({
+        selector: ".js_country_checkout_vat",
+        start: function () {
+            this._super();
```
Comment (yajo) [inline website_sale_checkout_country_vat/static/src/js/website_sale_checkout_country_vat.js]: This whole file does nothing AFAICS.

File: website_sale_checkout_country_vat/views/assets.xml [hunk 13]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- © 2016 Sergio Teruel <sergio.teruel@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
```
Comment (yajo) [inline website_sale_checkout_country_vat/views/assets.xml]: LGPL

File: website_sale_checkout_country_vat/views/res_config.xml [hunk 14]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- © 2016 Sergio Teruel <sergio.teruel@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl-3). -->
```
Comment (yajo) [inline website_sale_checkout_country_vat/views/res_config.xml]: LGPL

File: website_sale_checkout_country_vat/views/templates.xml [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+<template id="country_dropdown"
+          inherit_id="website_snippet_country_dropdown.country_dropdown"
+          primary="True">
+
+    <xpath expr="//t[@t-set='default_country']" position="attributes">
+        <attribute name="t-value">website.vat_code_id</attribute>
+    </xpath>
+
+    <xpath expr="//input[@id='no_country_field']" position="attributes">
+        <attribute name="t-att-value">checkout.get('vat')</attribute>
```
Comment (yajo) [inline website_sale_checkout_country_vat/views/templates.xml]: This is wrong. You are setting a value for a reusable snippet that comes from a `checkout` variable, which is only available under some circumstances. That would make Odoo fail on any editor page I guess. You should use variables in your reusable template that you can override at `t-call` time, and then you will not need to replace anything from it.

Comment (yajo) [inline website_sale_checkout_country_vat/views/templates.xml]: Oh sorry, I missed that you were using `primary=True`. Anyway, the variables approach is much better.

File: website_sale_checkout_country_vat/views/templates.xml [hunk 16]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- © 2016 Sergio Teruel <sergio.teruel@tecnativa.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl-3). -->
```
Comment (yajo) [inline website_sale_checkout_country_vat/views/templates.xml]: LGPL

File: website_sale_checkout_country_vat/README.rst [hunk 17]
```diff
@@ -0,0 +1,68 @@
+.. image:: https://img.shields.io/badge/licence-LGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/lgpl-3.0-standalone.html
+   :alt: License: LGPL-3
+
+=================================
+Website Sale Checkout Country VAT
```
Comment (yajo) [inline website_sale_checkout_country_vat/README.rst:6]: Module should be called `website_sale_vat_country_dropdown` IMHO.

Comment (rafaelbn) [inline website_sale_checkout_country_vat/README.rst:6]: cc @pedrobaeza

**Comment by yajo** [comment]
Squashed some commits with little fixes, remove controllers, CSS, js, make use of variables instead of replacing the template.

:+1:

**Comment by rafaelbn** [comment]
Please @sergio-teruel could you rebase now that https://github.com/OCA/website/pull/210 is merged?

**Comment by yajo** [comment]
Done.

**Comment by rafaelbn** [comment]
Hi, it doesnt work. I set Spain and VAT code is other:

**Comment by pedrobaeza** [comment]
They are different modules, @rafaelbn. The field you are talking about is provided by https://github.com/OCA/e-commerce/tree/8.0/website_sale_default_country, not by standard Odoo, so there are 2 possibilities:
- There's a way to make compatible both without depending one in another
- There's the need of creating a glue module to close the behavior with all installed.

**Comment by rafaelbn** [comment]
I read README: https://github.com/OCA/e-commerce/pull/110/files#diff-a0d0ee1d751a4c37098995c2952d248dR17

**Comment by pedrobaeza** [comment]
Yeah, sorry, I didn't see that new option. But it chokes with the other module, so we should check with the developers.

**Comment by yajo** [comment]
Hi, @rafaelbn is right. I'm removing that option since with the new defaults in https://github.com/OCA/website/pull/223, you can just change the `country_id` for the Template/Public users and it has the same effect. I'm also adding a default here to use the `country_id` from the checkout context itself. I'll push soon.

**Comment by pedrobaeza** [comment]
I've tried all possible scenarios for the country default:
- Public user. No country pre-selected.
- Public user. Default country. Default country selected.
- Public user. Change country in form. Country in the dropdown changed.
- Registered user with country. Country selected in checkout by default.

So it's a :+1:

Thanks for the excellent work! Merging

**Pull request merged**